cmd/dp-observation-extractor/main.go

```diff
@@ -33,29 +56,84 @@ func main() {
 		config.FileConsumerTopic,
 		config.FileConsumerGroup,
 		kafka.OffsetNewest)
-	if err != nil {
-		log.Error(err, log.Data{"message": "failed to create kafka consumer"})
-		os.Exit(1)
-	}
+	checkForError(err)
 
-	kafkaProducer := kafka.NewProducer(config.KafkaAddr, config.ObservationProducerTopic, 0)
+	kafkaObservationProducer, err := kafka.NewProducer(config.KafkaAddr, config.ObservationProducerTopic, 0)
+	checkForError(err)
 
-	signals := make(chan os.Signal, 1)
-	signal.Notify(signals, os.Interrupt, syscall.SIGTERM)
+	kafkaErrorProducer, err := kafka.NewProducer(config.KafkaAddr, config.ErrorProducerTopic, 0)
+	checkForError(err)
 
-	go func() {
-		<-signals
+	observationWriter := observation.NewMessageWriter(kafkaObservationProducer)
+	eventHandler := event.NewCSVHandler(s3, observationWriter)
+
+	eventConsumer := event.NewConsumer()
+	eventConsumer.Consume(kafkaConsumer, eventHandler)
+
+	shutdownGracefully := func() {
+
+		ctx, cancel := context.WithTimeout(context.Background(), config.GracefulShutdownTimeout)
 
 		// gracefully dispose resources
-		kafkaConsumer.Closer() <- true
-		kafkaProducer.Closer() <- true
+		err = eventConsumer.Close(ctx)
+		if err != nil {
+			log.Error(err, nil)
+		}
+
+		err = kafkaConsumer.Close(ctx)
+		if err != nil {
+			log.Error(err, nil)
+		}
+
+		err = kafkaErrorProducer.Close(ctx)
+		if err != nil {
+			log.Error(err, nil)
+		}
+
+		err = kafkaObservationProducer.Close(ctx)
+		if err != nil {
+			log.Error(err, nil)
+		}
+
+		err = httpServer.Shutdown(ctx)
+		if err != nil {
+			log.Error(err, nil)
+		}
+
+		// cancel the timer in the shutdown context.
+		cancel()
 
 		log.Debug("graceful shutdown was successful", nil)
 		os.Exit(0)
```

Comment:
shouldn't this be non-zero, indicating some failure?

Reply:
Yes - have updated it 👍